Migrate test_get_next_results to Travis suite

davehunt opened the pull request:
This is currently failing for me locally as it's always getting one additional result set than expected. Could you take a look @edmorley and see if I'm missing anything obvious? Note that I've left in some debug so we're sleeping 5 seconds instead of waiting. This gives a more informative assertion error as it shows how the count does not match expected, rather than timing out waiting for it to match.

edmorley commented on tests/selenium/test_get_next_results.py:
```diff
@@ -0,0 +1,30 @@
+
+
+@pytest.mark.parametrize('count', ((10), (20), (50)))
+def test_get_next_results(base_url, selenium, count, push_stored):
```
Should `push_stored` be `sample_push`?

davehunt replied:
I need it to be stored though. I could overwrite the `push_stored` fixture myself, but this seemed cleaner.

edmorley replied:
Oh you can override fixtures by declaring them locally in pytest, I didn't realise. Ah so `push_stored` uses the updated `sample_push`, makes sense.

edmorley commented:
Hmm I'm not 100% sure, but maybe it's due to this?
https://github.com/mozilla/treeherder/blob/28a2ff5cb8007574c1e2edcbc9071d699424cd14/ui/js/models/resultset.js#L67-L70

@camd ?

davehunt commented:
I realised I was creating pushes in the future, but I've since tried changing each push to be in the past. I even tried the first push to be 1 day ago and all subsequent pushes to be spaced a day apart, and there was no difference to the test result.

edmorley commented:
If the revision is set to a 40 character random hex value, does that help? (The query does a LIKE internally to help support 12 character SHAs, perhaps several of the faked revisions are overlapping - eg `revision-1` and `revision-11`)

davehunt commented:
Same result with randomly generated strings of 40 characters. Always one more push than expected.

davehunt commented:
Pinging @camd to see if he has some time to look into this.

camd commented:
@davehunt and @edmorley : Ed, you're right, this is the cause of getting 11.  The intent here was to get an overlap of one with the oldest resultset we have.  The reason was that we need to get the next 10 based on the ``push_timestamp``.  Here's the scenario we were trying to handle:

* If (and I'm not sure this is possible, tbh) there were 2 pushes with the exact same ``push_timestamp``.  
* In our current list of pushes, the last (oldest) one is just ONE of those two with the same timestamp
* we fetch the next 10 pushes with ``push_timestamp__lt`` (less than)

In this scenario, we'd miss the second of the pushes with identical timestamp.  The way we have it written here, you'd get an overlap of the oldest push, which we'd ignore, since we already have it.

This is one of those cases of, "It seemed like a good idea at the time" and, as I mentioned, I'm not sure two pushes could have identical ``push_timestamp``s.  That assumption may have been based on the concept that ``resultset``s could be groupings of commits from different repos vs. ``push``es only coming from a single repo.  Now that we only have pushes, that assumption could well be false and we could change the query to ``push_timestamp_lt`` and use the real count.

davehunt commented:
I can confirm that removing the following causes the test to pass:

https://github.com/mozilla/treeherder/blob/28a2ff5cb8007574c1e2edcbc9071d699424cd14/ui/js/models/resultset.js#L67-L70

I've experimented with more pushes, sorting the pushes by timestamp (ascending and descending), making all pushes exist in the past, having pushes at least an hour or day apart, and nothing I've tried fixes this. Looking at the code I don't actually understand why this isn't failing in production...

Any help would be appreciated @edmorley and @camd. This is now the final test to migrate! 😁

davehunt commented:
I've rebased this and removed Jenkins directory entirely as this is now the last test to migrate. It's still failing, and I'm unable to confirm that this isn't a bug. Hopefully I can get some assistance from @camd or @edmorley to push this through, otherwise it may make sense to relax the assertion or to mark the test as expected to fail.

edmorley commented:
I'm out of ideas for this one, but hopefully @camd might know.

Could you also remove the jenkins directory references from `setup.cfg`? I guess as well we can now mark the treeherder repository as `onhold` in `treeherder/model/fixtures/repository.json`?

davehunt commented:
> Could you also remove the jenkins directory references from `setup.cfg`?

Done

> I guess as well we can now mark the treeherder repository as `onhold` in `treeherder/model/fixtures/repository.json`?

Done

camd commented:
I've been working on this for a few hours today and it's still mysterious so far to me.  @davehunt, I'll try pinging you tomorrow for a little help.  But my two main questions are:

1. Is there a way to have console output exposed when running the test?  I'd like to output the URL that is being used to fetch the "next N pushes"
2. Is there a way to pause things and open it in a normal browser to try to verify what's going on?

camd commented on tests/selenium/test_get_next_results.py:
```diff
@@ -0,0 +1,30 @@
+    for i in range(61):
+        sample_push.append({
+            'revision': 'revision{}'.format(i),
+            'push_timestamp': time.time() + (i * 3600),
```
In reality, you could probably also just use ``42`` instead of the ``time.time()``.   The base doesn't make a difference.

davehunt commented:
Thank you! I've updated and pushed. Have a great PTO @camd!
